state/protocol/snapshot.go

```diff
@@ -49,22 +49,25 @@ type Snapshot interface {
 	// as of this block. It represents the sealed state.
 	Commit() (flow.StateCommitment, error)
 
-	// SealingSegment returns the chain segment such that the head (greatest
-	// height) is this snapshot's reference block and the tail (least height)
+	// SealingSegment returns the chain segment such that the highest block
+	// is this snapshot's reference block and the lowest block
 	// is the most recently sealed block as of this snapshot (ie. the block
 	// referenced by LatestSeal). The segment is in ascending height order.
 	//
 	// TAIL <- B1 <- ... <- BN <- HEAD
 	//
-	// NOTE 1: TAIL is not always sealed by HEAD. In the case that the head of
-	// the snapshot contains no seals, TAIL must be sealed by the first ancestor
-	// of HEAD which contains any seal.
+	// NOTE 1: The highest block in the sealing segment must have a seal
+	// for the lowest block in the sealing segment.
```

Comment:
Thinking about this more, I don't think we can use this stricter constraint.

Consider the chain segment, where B contains the execution receipt for A and C seals A. (D does not contain a seal.)
```
A<-B(R_A)<-C(S_A)<-D
```

We would like for every queryable protocol snapshot to have a valid sealing segment. I should be able to do `state.AtBlockID(C.ID())` and `state.AtBlockID(D.ID())` and receive a snapshot with a valid sealing segment for both. 

But since D does not contain a seal, it does not satisfy the stricter requirement we're introducing here.
> The highest block in the sealing segment must have a seal for the lowest block in the sealing segment.

I suggest we:
* revert this constraint on the sealing segment documentation to the previous definition: "Either HIGHEST seals LOWEST, or the first ancestor of HIGHEST containing any seal, seals LOWEST"
* make sure sealing segment building and validation allows for the case where the highest block contains no seal

Reply:
Interesting , in this case would this [line]( https://github.com/onflow/flow-go/blob/master/state/protocol/badger/snapshot.go#L269) not return an error because D does not contain a seal ?

Reply:
It won't. We index the _latest_ seal each time we insert a block ([reference](https://github.com/onflow/flow-go/blob/master/state/protocol/badger/mutator.go#L439-L443)), so `seals.ByBlockID` will return a seal for all inserted block IDs, even if the block did not contain any seal.

Reply:
Ahh okay that makes sense, I will update hasValidSeal to reflect this and revert the godoc

Reply:
> 	// NOTE 1: TAIL is not always sealed by HEAD. In the case that the head of
> 	// the snapshot contains no seals, TAIL must be sealed by the first ancestor
> 	// of HEAD which contains any seal.

In the case where D **_DOES_** contains seals but **_DOES NOT_** contain a seal for A do we still check for a valid ancestor or do we consider this invalid state?

Reply:
Exactly, If we see a sealing segment like `A<-B<-C<-D`, and D has no seal for A, but does have a different seal (eg. for B), then that is an invalid sealing segment.

Reply:
https://github.com/onflow/flow-go/pull/1533/commits/5ac9897faba9fa9e54bf8482bac9135677167f68